pyth-rng/src/config.rs

```diff
@@ -90,3 +95,60 @@ pub struct RandomnessOptions {
     #[arg(default_value = "32")]
     pub chain_length: u64,
 }
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct Config {
+    pub chains: Vec<EthereumConfig>,
+}
+
+impl Config {
+    pub fn check_is_valid(&self) -> Result<(), Box<dyn Error>> {
+        let counts: HashMap<_, _> = self.chains.iter().fold(HashMap::new(), |mut map, s| {
+            *map.entry(s.chain_id.clone()).or_insert(0) += 1;
+            map
+        });
+
+        let duplicates: Vec<_> = counts
+            .iter()
+            .filter(|&(_, &c)| c > 1)
+            .map(|(id, _)| id.clone())
+            .collect();
+
+        if duplicates.len() > 0 {
+            Err(anyhow!(format!(
+                "Config has duplicated chain ids: {}",
+                duplicates.join(",")
+            ))
+            .into())
+        } else {
+            Ok(())
+        }
+    }
```

Comment:
It would be nice to incorporate this into the parsing to avoid the decoupled check code, ala [dont validate, parse](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/). Should be able to do this with something that looks like this:


```rust
mod chains {
    pub fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<HashMap<ChainId, String>, D::Error> {
        let vec: Vec<String> = Vec::deserialize(deserializer)?;
        let mut map = HashMap::new();
        ensure!(vec.duplicates_by(|v| v.chain_id).count() == 0, "duplicate chains in config");
        vec.into_iter()
            .try_fold(
                HashMap::new(),
                |mut acc, item| acc.insert(item.chain_id, item)
            )
    }
}

#[derive(Clone, Debug, serde::Deserialize)]
pub struct Config {
    #[serde(with = "chains")]
    pub chains: Vec<EthereumConfig>,
}
```

Note that along with the other suggestion of moving `load` to `from_file` cleans this up a lot.

Reply:
ah, smart. I ended up using the default serde HashMap support (didn't realize that was in there), but this is much better.

I didn't extend the chain config to an enum, as i think this is easy to do later when we need it.